module/preference/PreferenceDialog.kt

```diff
@@ -0,0 +1,500 @@
+import com.vaticle.typedb.studio.state.common.util.Label.ENABLE_GRAPH_OUTPUT
+import com.vaticle.typedb.studio.state.common.util.Label.GRAPH_VISUALISER
+import com.vaticle.typedb.studio.state.common.util.Label.MANAGE_PREFERENCES
+import com.vaticle.typedb.studio.state.common.util.Label.OK
+import com.vaticle.typedb.studio.state.common.util.Label.PROJECT_IGNORED_PATHS
+import com.vaticle.typedb.studio.state.common.util.Label.PROJECT_MANAGER
+import com.vaticle.typedb.studio.state.common.util.Label.QUERY_RUNNER
+import com.vaticle.typedb.studio.state.common.util.Label.RESET
+import com.vaticle.typedb.studio.state.common.util.Label.SET_QUERY_LIMIT
+import com.vaticle.typedb.studio.state.common.util.Label.TEXT_EDITOR
+import com.vaticle.typedb.studio.state.common.util.Message.Preference.Companion.UNEXPECTED_ERROR
+import com.vaticle.typedb.studio.state.common.util.Sentence.IGNORED_PATHS_CAPTION
+import com.vaticle.typedb.studio.state.common.util.Sentence.PREFERENCES_GRAPH_OUTPUT_CAPTION
+import com.vaticle.typedb.studio.state.common.util.Sentence.PREFERENCES_MATCH_QUERY_LIMIT_CAPTION
+import com.vaticle.typedb.studio.state.page.Navigable
+import mu.KotlinLogging
+
+object PreferenceDialog {
+    private val WIDTH = 800.dp
+    private val HEIGHT = 600.dp
+    private val NAVIGATOR_INIT_SIZE = 200.dp
+    private val NAVIGATOR_MIN_SIZE = 150.dp
+    private val PREFERENCE_GROUP_INIT_SIZE = 600.dp
+    private val PREFERENCE_GROUP_MIN_SIZE = 500.dp
+    private val RESET_BUTTON_HEIGHT = 20.dp
+
+    private val preferenceMgr = StudioState.preference
+    private val notificationMgr = StudioState.notification
+    private var focusedPreferenceGroup by mutableStateOf(PreferenceGroup(""))
+    private var state by mutableStateOf(PreferencesForm())
+
+    private val LOGGER = KotlinLogging.logger {}
+
+    sealed interface PreferenceField {
+        val label: String
+        val caption: String?
+        fun isValid(): Boolean
+        @Composable fun Display()
+
+        @Composable
+        fun Layout(fieldContent: @Composable () -> Unit) {
+            Field(label) {
+                fieldContent()
+            }
+            if (!caption.isNullOrBlank()) {
+                Caption()
+            }
+        }
+
+        @Composable
+        fun Caption() {
+            CaptionText(caption!!)
+        }
+
+        class TextInputValidated(initialValue: String,
+                                 override val label: String, override val caption: String? = null,
+                                 private val placeholder: String, private val invalidWarning: String,
+                                 private val validator: (String) -> Boolean = { true }) : PreferenceField {
+
+            var value by mutableStateOf(initialValue)
+
+            @Composable
+            override fun Display() {
+                val positionProvider = rememberComponentRectPositionProvider(
+                    offset = DpOffset(0.dp, -(2 * Form.FIELD_HEIGHT.value).dp)
```

Comment:
We want our popup to appear above the `TextInput` field form, not below (covering up the caption) or on top of (covering up the current value).